### DISCUSSION TITLE
coreWCF as windows service in dotnet 6

### DISCUSSION THREAD

**Discussion by chuckthompsonprofisee** [CoreWCF/CoreWCF#703] [General]
Anyone have any examples of how to hows the coreWCV host in a windows service in dotnet 6.  
I started with the web app example, but in a service it uses IHost host = Host.CreateDefaultBuilder which does not allow UseServiceModel like it does in the webappbuilder.  
Anyone know how to get around this?

**Comment by davidfowl** [comment]
Can you show your code?

**Comment by chuckthompsonprofisee** [comment]
I followed this to make a windows service
https://docs.microsoft.com/en-us/dotnet/core/extensions/windows-service
and then I was trying to follow this to add the wcf service
https://devblogs.microsoft.com/dotnet/corewcf-v1-released/

but with the windows service there is no builder like this:

```C#
var builder = WebApplication.CreateBuilder(args);
builder.WebHost.ConfigureKestrel((context, options) =>
{
options.AllowSynchronousIO = true;
});

// Add WSDL support
builder.Services.AddServiceModelServices().AddServiceModelMetadata();
builder.Services.AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();

var app = builder.Build();
app.UseServiceModel(builder =>
{
builder.AddService((serviceOptions) => { })
// Add a BasicHttpBinding at a specific endpoint
.AddServiceEndpoint<EchoService, IEchoService>(new BasicHttpBinding(), "/EchoService/basichttp")
// Add a WSHttpBinding with Transport Security for TLS
.AddServiceEndpoint<EchoService, IEchoService>(new WSHttpBinding(SecurityMode.Transport), "/EchoService/WSHttps");
});
var serviceMetadataBehavior = app.Services.GetRequiredService();
serviceMetadataBehavior.HttpGetEnabled = true;

app.Run();
```

what I have is this:

```C#
using IHost host = Host.CreateDefaultBuilder(args)
    .UseWindowsService(options =>
    {
        options.ServiceName = ".NET Joke Service";
    })
    .ConfigureServices(services =>
    {
        services.AddSingleton<JokeService>();
        services.AddHostedService<WindowsBackgroundService>();
    })
    .Build();

await host.RunAsync();
```

and I cant add the .UseServiceModel() as its not an option like it is to call .ConfigureServices()

**Reply by mconnew** [reply]
Sorry for the (really) late response. The sample you are following is for when your service is NOT a web service. aspnetcore has a concept of a hosted service, which is basically just a class which implements IHostedService which has a lifetime where it can be started and stopped, and take part in the whole DI ecosystem of aspnetcore. Some of the transports in CoreWCF are actually implemented as an IHostedService, but that's an implementation detail you don't need to be concerned about.  
The extension method `UseWindowsService` extends `IHostBuilder`, but there's another method you can call against an `IServiceCollection` instead. You can use `AddWindowsService` when your app is a Web app. I haven't typed this into an IDE and tried to compile it so might have some errors, but something like this should work:
```csharp
var builder = WebApplication.CreateBuilder(args);
builder.WebHost.ConfigureKestrel((context, options) =>
{
    options.AllowSynchronousIO = true; // Note only needed now if using Streamed transfer mode, can probably remove
});

// Add WSDL support
builder.Services.AddServiceModelServices().AddServiceModelMetadata();
builder.Services.AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();

// Add Windows Service support
builder.Services.AddWindowsService(options =>
{
    options.ServiceName = ".NET Joke Service";
});

var app = builder.Build();
app.UseServiceModel(builder =>
{
    builder.AddService<EchoService>((serviceOptions) => { })
    // Add a BasicHttpBinding at a specific endpoint
    .AddServiceEndpoint<EchoService, IEchoService>(new BasicHttpBinding(), "/EchoService/basichttp")
    // Add a WSHttpBinding with Transport Security for TLS
    .AddServiceEndpoint<EchoService, IEchoService>(new WSHttpBinding(SecurityMode.Transport), "/EchoService/WSHttps");
});
var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
serviceMetadataBehavior.HttpGetEnabled = true;
app.Run();
```

**Comment by ghost** [comment]
have similar issue. trying to host coreWCF as windows service without IIS. trying with workerservice template. **ServiceHost** is not resolving

<Project Sdk="Microsoft.NET.Sdk.Worker">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <UserSecretsId>dotnet-WorkerService1-8f6007a8-3cd6-4f45-94e4-27bd6ffc03c3</UserSecretsId>
  </PropertyGroup>
  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="CoreWCF.Http" Version="1.5.1" />
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="6.0.1" />
  </ItemGroup>
</Project>


 protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 {
     while (!stoppingToken.IsCancellationRequested)
     {
         _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);
         await Task.Delay(1000, stoppingToken);
     }

     using (var host = new ServiceHost(typeof(MyWcfService)))
     {
         // Configure bindings, behaviors, and other settings for your WCF service
         var endpoint = host.AddServiceEndpoint(typeof(IMyWcfService), new BasicHttpBinding(), "http://localhost:8000/MyWcfService");

         try
         {
             // Open the service host
             host.Open();

             // Log that the service has started
             Console.WriteLine("CoreWCF service is running...");

             // Keep the background service alive until cancellation is requested
             while (!stoppingToken.IsCancellationRequested)
             {
                 await Task.Delay(1000, stoppingToken);
             }
         }
         catch (Exception ex)
         {
             // Handle exceptions
             Console.WriteLine($"An error occurred: {ex.Message}");
         }
         finally
         {
             // Close the service host when stopping
             host.Close();
         }
     }
 }

**Comment by ShadyHarbinger** [comment]
Can anyone run a .NET 5-8 CoreWCF application as Windows Service? I migrated one WCF service to CoreWCF .NET 8. It turned out that it is useless because cannot run as Windows Service. :(

**Reply by mconnew** [reply]
Does [my response here](https://github.com/CoreWCF/CoreWCF/discussions/703#discussioncomment-10574248) help you?

**Reply by ShadyHarbinger** [reply]
Sorry for the late response. For some reason, I have just got the notification in my mailbox. Yeah. It helped. I was so dumb I overlooked a missing/wrong using. This is why the extension method was not recognized and I was not sure that such an extension method exists. I am ashamed :D Thank you for the hard work on CoreWCF.

**Reply by rumel36** [reply]
> Sorry for the late response. For some reason, I have just got the notification in my mailbox. Yeah. It helped. I was so dumb I overlooked a missing/wrong using. This is why the extension method was not recognized and I was not sure that such an extension method exists. I am ashamed :D Thank you for the hard work on CoreWCF.

Did you use any windows service separately to host the CoreWCF?

**Comment by rumel36** [comment]
@chuckthompsonprofisee , Did you solve this issue? I am facing the same.

**Reply by sharpjoe** [reply]
Is there a plan to add the .UseNetTcp support into the stuff that works with IHostBuilder... struggling with a way to make a ported basic web service work over net.tcp in a windows service.  For a variety of reasons and draconian policies at this customer I cannot use an HTTP transport on the port I need it on.   It actually fires up but without the UseNetTcp I can't find way to force it to listen on IPv6 without putting IP's in the config.

**Reply by mconnew** [reply]
For now you can use `WebApplicationBuilder`. That has a `Host` property which is of type `ConfigureHostBuilder` that implements `IHostBuilder`. This enables you to use `UseWindowsService()` to make your app a Windows service. It also has a `WebHost` property, which is of type `ConfigureWebHostBuilder` which implements `IWebHostBuilder` and enables you to call `UseNetTcp`. You get the best of both worlds.
